src/libutil/position.cc

```diff
@@ -110,4 +110,50 @@ void Pos::LinesIterator::bump(bool atFirst)
     input.remove_prefix(eol);
 }
 
+std::string Pos::getSnippetUpTo(const Pos & end) const {
```

Comment:
Don't know what's the rationale behind this?

Maybe just parse the doc-comment range with yacc?

Reply:
This lets us avoid storing copies of the comments in memory.

Reply:
AFAIK the `Pos` contains "source" information, I think it's a bit odd to have a range betwern `(origin1, position1)` and `(origin2, position2)`, what will happen if `origin1 != origin2`? (Even if we have assert statement to ensure this will not happen, but the code is "semantically" weired.


 if indeed we don't want to copy strings how about `std::string_view`s with a dedicated string storage?

Reply:
It is the exact range which contains the comment. I suppose we could store only the start position, but then we'd have to re-parse the comment to find the end. `PosIdx` is 4 bytes, so we basically get to store the end for free when used in an 8 byte aligned situation on 64 bit machines.

You're right that there's a redundancy because of the extra origin, but I don't think removing that redundancy would make a substantial difference, and it might make this method harder to use.


> how about `std::string_view`s with a dedicated string storage?

That could also work, but I don't think we're already storing all file contents in memory and I'm not sure we should.

Reply:
> That could also work, but I don't think we're already storing all file contents in memory 

Currently we not. `getSource` works with "reading" the file path again (as per my understanding of this codebase).

In my previous career in clang there are `SourceManager` to store all file contents in memory, that will not consume much (~10MB for a whole nixpkgs). And allocate `PosIdx`-like handler to reference locations in a source. Accessing files from a filesystem is somehow *not* very good (just my personal preferences, because I feel there maybe `lazytrees` to allowing "in-memory" filesystem for nix in the future and current solution  may have conflicts about that).

> and I'm not sure we should.

I think we should start with a carefully revised version for this feature, otherwise we'll ending up having a lot of workarounds about it (e.g. how about `lazytrees`?).


Also current implementation does not handle non-ASCII characters right?

Reply:
We've already merged `SourcePath` from `lazy-trees` (in 2.19 iirc?) so that's not a risk anymore.

> that will not consume much (~10MB for a whole nixpkgs).

I feel like it'd be more, but you may well be right and that's a very manageable amount of overhead if it's worth the benefits. I'm sure there's more benefits to such an architecture than changing `DocComment` to the equivalent of `string_view`, but I don't quite see what they would be, and I think that deserves its own PR. That PR can show its quality by removing this code.